[IMP] Charts: Add Heatmap

anhe-odoo opened the pull request:
WIP1 - Add heatmap as new chart

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/o-spreadsheet/pull/4693.png)](https://mergebot.odoo.com/odoo/o-spreadsheet/pull/4693)

anhe-odoo committed:
[IMP] Charts: add heat map

Task Description

This PR aims to add a heatmap as a new chart type. The user has to
select 3 columns (x values/coordinates, y values/coordinates and z
values) as horizontal, vertical et data values. The chart uses the
same logic as calendar chart, and the color map and drawing part
has been refactored to use common code.

Related Task

Task: 6357421

anhe-odoo commented on src/helpers/figures/charts/runtime/chartjs_dataset.ts:
```diff
@@ -162,8 +162,8 @@ function getBarChartTotalLineDataset(
 
-export function getCalendarChartDatasetAndLabels(
-  definition: CalendarChartDefinition,
+export function getColorScaleGridDatasetAndLabels(
```
Not sure with these names though I have no better idea ATM :/

pro-odoo commented on src/components/figures/chart/chartJs/chartjs_show_values_plugin.ts:
```diff
@@ -86,6 +86,7 @@ export const chartShowValuesPlugin: Plugin = {
         break;
       case "calendar":
+      case "heatmap":
         drawCalendarValues(chart, options);
```
`drawCalendarValues` We should find a better name now that it's not only calendar

pro-odoo commented on src/components/side_panel/chart/chart_type_picker/chart_type_picker.ts:
```diff
@@ -67,7 +67,8 @@ export class ChartTypePicker extends Component<SpreadsheetChildEnv> {
+        definition.type === "bubble" ||
+        definition.type === "heatmap")
     ) {
       // Scorecard and gauge don't have a data source but can still be converted to other types of charts
```
and bubble and heatmap...
It's maybe time to make a list, and check if the type is include in that list

hokolomopo commented on src/components/figures/chart/chartJs/chartjs_colorscale_plugin.ts:
```diff
@@ -76,6 +79,21 @@ export const chartColorScalePlugin: Plugin = {
     ctx.fillText(formattedMinValue, gradientX + gradientWidth + 5, gradientY + gradientHeight - 6);
     ctx.fillText(formattedMaxValue, gradientX + gradientWidth + 5, gradientY + 6);
 
+    // Draw a swatch for the missing-value color, with a label, below the gradient
+    if (options.missingValueColor) {
+      const swatchSize = gradientWidth;
+      const swatchY = gradientY + gradientHeight + 20;
```
... a swatch ? whass'that ?

hokolomopo commented on src/components/figures/chart/chartJs/chartjs_colorscale_plugin.ts:
```diff
@@ -76,6 +79,21 @@ export const chartColorScalePlugin: Plugin = {
+      ctx.lineWidth = 1;
+      ctx.strokeRect(gradientX, swatchY, swatchSize, swatchSize);
+
+      ctx.fillStyle = options.fontColor ?? "black";
```
default to chartJs default instead

hokolomopo commented on src/components/side_panel/chart/building_blocks/axis_design/axis_design_editor.ts:
```diff
@@ -38,6 +38,8 @@ export class AxisDesignEditor extends Component<SpreadsheetChildEnv> {
     definition: types.ChartWithAxisDefinition(),
     updateChart: types.function<ChartUpdateFunction>(),
     axesList: types.ArrayOf<AxisDefinition>(),
+    editBoundaries: types.boolean().optional(true),
+    editGridlines: types.boolean().optional(true),
```
```suggestion
    canEditBoundaries: types.boolean().optional(true),
    canEditGridlines: types.boolean().optional(true),
```
nitpick

hokolomopo commented on src/components/side_panel/chart/chart_type_picker/chart_previews.xml:
```diff
@@ -417,6 +417,25 @@
     </svg>
   </t>
   <t t-name="o-spreadsheet-ChartPreview.CALENDAR_CHART">
+    <svg class="o-chart-preview" viewBox="0 0 48 48">
```
It's pretty <3

hokolomopo commented on src/components/side_panel/chart/heatmap_chart/heatmap_chart_config_panel.ts:
```diff
@@ -0,0 +1,135 @@
+  static components = { ChartLabelRange, ChartErrorSection };
+  protected props = props(chartSidePanelPropsDefinition) as unknown as Props;
+
+  protected state: HeatmapChartPanelState = proxy({
+    rowRangeDispatchResult: undefined,
+    columnRangeDispatchResult: undefined,
+    dataRangeDispatchResult: undefined,
+  });
```
why do we have to store 3 dispatchResult ? can't we do a single one, the dispatch reasons are different anyway

hokolomopo commented on src/components/side_panel/chart/heatmap_chart/heatmap_chart_config_panel.ts:
```diff
@@ -0,0 +1,135 @@
+export class HeatmapChartConfigPanel extends Component<SpreadsheetChildEnv> {
+  static template = "o-spreadsheet-HeatmapChartConfigPanel";
+  static components = { ChartLabelRange, ChartErrorSection };
+  protected props = props(chartSidePanelPropsDefinition) as unknown as Props;
```
```suggestion
  protected props = useProps(chartSidePanelPropsDefinition) as unknown as Props;
```

hokolomopo commented on src/components/side_panel/chart/heatmap_chart/heatmap_chart_config_panel.xml:
```diff
@@ -0,0 +1,33 @@
+<templates>
+  <t t-name="o-spreadsheet-HeatmapChartConfigPanel">
+    <div>
+      <ChartLabelRange
```
we should rename this component sometime, it's not always used for labels now

hokolomopo commented on src/helpers/figures/charts/runtime/chart_custom_tooltip.ts:
```diff
@@ -12,28 +12,37 @@ const templates = /* xml */ `
                 t-attf-style="max-width: {{ labelsMaxWidth }}"
               />
+              <small
+                t-else=""
```
probably need a isArray, and that's why it crashed (idk why we had undefined labels tho)

hokolomopo commented on src/helpers/figures/charts/runtime/chart_custom_tooltip.ts:
```diff
@@ -12,28 +12,37 @@ const templates = /* xml */ `
-            <th class="o-tooltip-title align-baseline border-0 text-truncate" t-out="title" t-attf-style="max-width: {{ labelsMaxWidth }}"/>
-          </tr>
-        </thead>
+      <div
+        t-if="title"
+        class="o-tooltip-title text-truncate mb-1 fw-bold"
+        t-out="title"
+        t-attf-style="max-width: {{ labelsMaxWidth }}"
+      />
+      <table class="overflow-hidden m-0" style="table-layout: auto; width: auto">
```
not sure if using a table still make sense if you had to move the title out of the table. The HTML might be simpler with flexes

hokolomopo commented on src/helpers/figures/charts/runtime/chart_data_extractor.ts:
```diff
@@ -122,72 +123,99 @@ function getDateTimeLabel(value: number, stamp: CalendarChartGranularity): strin
+  const yValues: K[] = [];
+  const previousYValues: K[] = [];
+  for (let i = 0; i < count; i++) {
+    const xValue = getXKey(i);
+    const xKey = String(xValue);
```
getXKey doesn't return a key ?

hokolomopo commented on src/helpers/figures/charts/runtime/chart_data_extractor.ts:
```diff
@@ -122,72 +123,99 @@ function getDateTimeLabel(value: number, stamp: CalendarChartGranularity): strin
-      locale
-    );
-    if (!(xValue in grouping)) {
+  const xValues: K[] = [];
+  const yValues: K[] = [];
+  const previousYValues: K[] = [];
```
You should use sets instead of arrays. And I'm not sure what is previousYValues used for ?

hokolomopo commented on src/helpers/figures/charts/runtime/chart_data_extractor.ts:
```diff
@@ -122,72 +123,99 @@ function getDateTimeLabel(value: number, stamp: CalendarChartGranularity): strin
+ * calendar chart (x/y keys derived from date granularities) and the heatmap chart (x/y keys
+ * are raw category values from two separate ranges).
+ */
+function computeGroupedValuesAndLabels<K extends string | number>(
+  count: number,
+  getXKey: (i: number) => K,
+  getYKey: (i: number) => K,
+  getValue: (i: number) => FunctionResultObject,
+  formatXLabel: (key: K) => string,
+  formatYLabel: (key: K) => string,
+  sortXKeys: (keys: K[]) => K[],
+  sortYKeys: (keys: K[]) => K[],
+  countEntries: boolean = false
+): { dataSetsValues: DatasetValues[]; labels: LabelValues } {
```
I would name the arguments (passing a POJO as arg). Calls to this methods are very hard to read.

hokolomopo commented on src/helpers/figures/charts/runtime/chart_data_extractor.ts:
```diff
@@ -122,72 +123,99 @@ function getDateTimeLabel(value: number, stamp: CalendarChartGranularity): strin
   };
 }
 
+function computeValuesAndLabels(
```
```suggestion
function computeCalendarValuesAndLabels(
```

hokolomopo commented on src/helpers/figures/charts/runtime/chart_data_extractor.ts:
```diff
@@ -226,6 +254,157 @@ export function getCalendarChartData(
+ * indexes past the end of an array are treated as missing.
+ *
+ * When `dataValues` is `undefined` (no data range set), each valid (row, column) pair is given a
+ * value of 1 instead of requiring a numeric value: summed later by `computeGroupedValuesAndLabels`,
+ * this makes each heatmap cell show how many times that (row, column) pair occurs.
+ */
+function filterInvalidHeatmapDataPoints(
```
transformToValidHeatmapDataPoints maybe? filter doesn't seem correct if we generate an array of valid values on the fly

hokolomopo commented on src/helpers/figures/charts/runtime/chart_data_extractor.ts:
```diff
@@ -226,6 +254,157 @@ export function getCalendarChartData(
+}
+
+/**
+ * Splits a series of numeric values into equal-width bins covering their whole range, using
+ * Sturges' formula for the bin count (same heuristic as the column statistics histogram).
```
can you use the same helper as the histogram then ?

hokolomopo commented on src/helpers/figures/charts/runtime/chartjs_scales.ts:
```diff
@@ -106,30 +106,48 @@ export function getBarChartScales(
+            stepSize: 1,
+            autoSkip: false,
+            color: fontColor,
+            callback: (value) => yTicksAtBoundaries[Math.round(Number(value))] ?? "",
```
rounding the value and using it as an index ? something seems fishy here.

hokolomopo commented on src/helpers/figures/charts/runtime/chartjs_scales.ts:
```diff
@@ -106,30 +106,48 @@ export function getBarChartScales(
-export function getCalendarChartScales(
-  definition: GenericDefinition<BarChartDefinition>,
-  datasets: ChartDataset[]
+// A column-boundary axis can't just be the bar-index "x" axis switched to a linear type: Chart.js
+// infers each bar's position from its array index only on a category scale, and a bar on a linear
+// index axis is centered on its x value rather than spanning [i, i+1] — both fragile to get
+// pixel-perfect. Instead, the real "x" axis stays a normal (hidden) category axis, exactly as
+// before, and a separate overlay axis — sharing the same "top" position and the same [0, binCount]
+// domain, but with no dataset bound to it — draws the boundary ticks on top of it.
+const X_BOUNDARY_AXIS_ID = "xBoundary";
```
that's a very cool AI generated docstring, and I understood none of it.

hokolomopo commented on src/helpers/figures/charts/runtime/chartjs_scales.ts:
```diff
@@ -142,16 +160,38 @@ export function getCalendarChartScales(
+      position: "top",
+      min: 0,
+      max: xTicksAtBoundaries.length - 1,
+      // Without this, a linear scale defaults to reserving half a step of margin on each side
+      // (meant for a scatter/line-style axis where edge points shouldn't be clipped), which
+      // insets its [0, binCount] domain from the chartArea and desyncs it from the category axis
+      // below, whose bars fill the chartArea edge to edge.
```
Why do we need two types of scale at all btw ? It just seem to complexify everything. Can't you just use the same scale type than the calendar charts ?  You already compute the bins by hand, so there should be no need to have chartJs do anything, and you can just give it the tick values, no ?

hokolomopo commented on src/helpers/figures/charts/runtime/chartjs_tooltip.ts:
```diff
@@ -419,10 +425,10 @@ function customTooltipHandler({ chart, tooltip }: ChartContext) {
 
   const tooltipItems = tooltip.body.map((body, index) => {
-    let label = body.before[0];
+    let label: string | string[] = body.before.length > 1 ? body.before : body.before[0];
```
can you make labels always an array ? do you have to do a special case when lenght === 1 ?

hokolomopo commented on src/helpers/figures/charts/heatmap_chart.ts:
```diff
@@ -0,0 +1,201 @@
+import { getChartTitle } from "./runtime/chartjs_title";
+import { getColorGridChartTooltip } from "./runtime/chartjs_tooltip";
+
+function checkRanges(definition: HeatmapChartDefinition<string>): CommandResult {
```
can be inlined, only used once

hokolomopo commented on src/helpers/figures/charts/heatmap_chart.ts:
```diff
@@ -0,0 +1,201 @@
+  sequence: 120,
+  allowedDefinitionKeys: [
+    ...AbstractChart.commonKeys,
+    "rowRange",
+    "columnRange",
+    "dataRange",
```
can you try to make something in common with bubble charts ? Intead of something specific to heatmaps. Maybe yRanges, xRange, sizeRange (sizeRange doesn't really fit, but it's just naming).
- would make it easier to convert between bubble/heatmap/calendar/other charts (eg. even if we only use one range in from the yRanges in a heatmap, we would have the same structure)
- would make it easier in the future to have a common handler for all of them, and integrate them in the data source system

hokolomopo commented on src/helpers/figures/charts/heatmap_chart.ts:
```diff
@@ -0,0 +1,201 @@
+    const { labels, datasets } = getColorGridChartDatasetAndLabels(definition, chartData);
+
+    const config: ChartConfiguration<"calendar"> = {
+      // Reuses the "calendar" chart.js controller: a thin bar-based grid renderer with no
+      // calendar-specific logic, so the same rendering pipeline works for the heatmap chart.
+      // This is purely an internal chart.js detail; the persisted definition.type is "heatmap".
```
remove those AI comments, write something yourself if you want. `thin bar-based` thin bars ? where ? `persisted definition.type is "heatmap"` well yeah ? it's the whole point of our chart runtimes

hokolomopo commented on src/types/chart/chart.ts:
```diff
@@ -293,6 +297,8 @@ export interface ChartRuntimeGenerationArgs {
   axisType?: AxisType;
   topPadding?: number;
   background: Color;
+  /** For a heatmap axis binning a numeric range, the formatted value at each bin boundary (bin count + 1 labels) */
+  binBoundaryLabels?: { x?: string[]; y?: string[] };
```
shouldn't it be axisTickLabels or something ? We do'nt really care that those are bins

hokolomopo commented on src/types/chart/common_chart.ts:
```diff
@@ -27,3 +27,16 @@ export interface DataSourceChartDefinition<T extends string | Range = Range>
   readonly dataSource?: undefined; // doesn't use a data source. Explicitly declaring the key ensures that `dataSource` can be safely accessed on the `ChartDefinition` union without TypeScript errors.
 }
+
+/**
+ * Shape shared by charts rendered as a bar-based grid colored with a color scale
+ * (calendar, heatmap), used by their common runtime helpers.
+ */
```
AI, remove

hokolomopo commented on src/types/chart/heatmap_chart.ts:
```diff
@@ -0,0 +1,18 @@
+  readonly rowRange?: T;
+  readonly columnRange?: T;
+  readonly dataRange?: T;
+  /** If true, the first cell of each range is a header and is excluded from the data. */
+  readonly dataSetsHaveTitle: boolean;
```
so the same as every other type of chart ? why does this one deserve a docstring

hokolomopo commented on src/helpers/figures/charts/heatmap_chart.ts:
```diff
@@ -0,0 +1,201 @@
+      columnRange,
+      rowRange,
+      dataSetsHaveTitle: isDataSourceRange ? context.dataSource.dataSetsHaveTitle ?? false : false,
+      legendPosition: context.legendPosition === "right" ? "right" : "left",
```
I'd keep the legend position as it, and do like geo chart and convert it at runtime/panel display. Or keep that, but do something cleaner:
1) really annoying that changing lien chart => heatmap => line chart would move the legend
2) nothing in the typing/commands prevent you from setting the legend at top/bottom in the definition

hokolomopo commented on tests/test_helpers/commands_helpers.ts:
```diff
@@ -432,6 +433,40 @@ export function createCalendarChart(
+      columnRange: data.columnRange,
+      dataRange: data.dataRange,
+      dataSetsHaveTitle: data.dataSetsHaveTitle ?? false,
+      legendPosition: data.legendPosition || "top",
```
top isn't supposed to be used for heatmap

hokolomopo commented on tests/figures/chart/heatmap/heatmap_chart_panel_component.test.ts:
```diff
@@ -0,0 +1,62 @@
+
+    expect(fixture.querySelector(".o-chart")).toBeTruthy();
+
+    const labels = fixture.querySelectorAll(".o-data-labels");
```
nitpick: can query selector the inputs directly, and use toHaveValue

hokolomopo commented on tests/figures/chart/heatmap/heatmap_chart_panel_component.test.ts:
```diff
@@ -0,0 +1,62 @@
+    await openChartConfigSidePanel(model, env, chartId);
+
+    // dataRange is "C2:C5": a vertical (multi-row) range -> header is the first row, row 2
+    expect(fixture.querySelector(".o-checkbox span")?.textContent).toBe("Use row 2 as headers");
```
toHaveText

hokolomopo commented on tests/figures/chart/heatmap/heatmap_chart_panel_component.test.ts:
```diff
@@ -0,0 +1,62 @@
+
+    // dataRange is "C2:C5": a vertical (multi-row) range -> header is the first row, row 2
+    expect(fixture.querySelector(".o-checkbox span")?.textContent).toBe("Use row 2 as headers");
+  });
```
maybe also a test that we can edit the various ranges ?

hokolomopo commented on tests/figures/chart/heatmap/heatmap_chart_panel_component.test.ts:
```diff
@@ -0,0 +1,62 @@
+});
+
+describe("Heatmap chart design side panel", () => {
+  test("Can change the missing value color", async () => {
+    const model = new Model({});
+    const { fixture, env } = await mountComponentWithPortalTarget(SidePanels, { model });
+    createHeatmapChart(model, TEST_CHART_DATA.heatmap, chartId, undefined, {});
+    await openChartDesignSidePanel(model, env, fixture, chartId);
+
+    const colorPickerButton = fixture.querySelector(
+      ".o-missing-value .o-round-color-picker-button"
+    );
+    expect(colorPickerButton).toBeTruthy();
+
+    const definition = model.getters.getChartDefinition(chartId) as HeatmapChartDefinition;
+    expect(definition.missingValueColor).toBeUndefined();
+  });
```
this doesn't test anything ???

hokolomopo commented on tests/figures/chart/heatmap/heatmap_chart_plugin.test.ts:
```diff
@@ -0,0 +1,459 @@
+    });
+  });
+
+  test("with a label range and 2 data series, uses them as (columnRange, rowRange, dataRange)", () => {
```
doesn't make much sense. Why would a random range of the datasets be treated totally differently from the others ? We don't do that for bubble charts. 

We should keep dataRange empty when converting from a chart with only 2 ranges types IMO.

hokolomopo commented on tests/figures/chart/heatmap/heatmap_chart_plugin.test.ts:
```diff
@@ -0,0 +1,459 @@
+    });
+  });
+
+  test("with no label range and 3 data series, uses them as (columnRange, rowRange, dataRange)", () => {
```
same here, but even worse. Let's NOT do magic

hokolomopo commented on tests/figures/chart/heatmap/heatmap_chart_plugin.test.ts:
```diff
@@ -0,0 +1,459 @@
+
+    // columns and rows keep the first-occurrence order of the source ranges, not sorted
+    expect(config.data.labels).toEqual(["Mon", "Tue", "Wed"]);
+    const datasets = config.data.datasets as any[];
+    expect(datasets.map((d) => d.label)).toEqual(["Apple", "Banana"]);
+    expect(datasets[0].values).toEqual([11, 2, undefined]); // Apple/Mon = 1 + 10
+    expect(datasets[1].values).toEqual([3, undefined, 4]);
```
nitpick: a single expect(dataset)toMatchObject might be more readable with label/values being in the same POJO (and avoid a casting as any)

Same in the other tests

hokolomopo commented on tests/figures/chart/heatmap/heatmap_chart_plugin.test.ts:
```diff
@@ -0,0 +1,459 @@
+      A3: "1",  B3: "1",   C3: "2",
+      A4: "10", B4: "100", C4: "30",
+    });
+    setCellFormat(model, "A2", "0%");
+    setCellFormat(model, "A3", "0%");
+    setCellFormat(model, "A4", "0%");
```
nitpick: we can set the format on the whole range in one command

hokolomopo commented on tests/figures/chart/heatmap/heatmap_chart_plugin.test.ts:
```diff
@@ -0,0 +1,459 @@
+    expect(colorScaleLegend.missingValueColor).toBeUndefined();
+  });
+
+  test("tooltip shows 'No Data' for a missing (row, column) combination", () => {
```
There should be no tooltip at all when there is no value ?

hokolomopo commented on tests/figures/chart/heatmap/heatmap_chart_plugin.test.ts:
```diff
@@ -0,0 +1,459 @@
+    expect(model.getters.getChartDefinition("1")).toMatchObject({ dataRange: "C2:C3" });
+  });
+
+  test("getDefinitionForExcel returns undefined, heatmap is not exportable to Excel", () => {
+    expect(HeatmapChart.getDefinitionForExcel({} as any, {} as any, {} as any)).toBeUndefined();
+  });
```
uh ? what in the casting hell is this ? no definition, the getters being an empty pojo casted as any ? It's lucky it doesn't just crash.

hokolomopo reviewed (commented):
👋 

I also added testing notes in the pad.

When using AI, please be better at re-reading and re-writing the output. It's not great...
